Forms/DefaultFormBuilder.php

```diff
@@ -0,0 +1,178 @@
+ * - updateFormFields
+ * - updateForm
+ */
+class DefaultFormBuilder implements FormBuilder {
+	use Extensible;
+
+	/**
+	 * @var Validator $validator
+	 */
+	protected $validator;
+
+	/**
+	 * @var Controller $controller
+	 */
+	protected $controller;
+
+	/**
+	 * @var DataObject $record
+	 */
+	protected $record;
+
+	/**
+	 * @param Controller $controller
+	 * @param DataObject $model
+	 * @param Validator $validator
+	 */
+	public function __construct($controller, $model, $validator = null) {
+		$this->setController($controller);
+		$this->setRecord($model);
+		if ($validator) {
+			$this->setFormValidator($validator);
+		}
+		$this->constructExtensions();
+	}
+
+	/**
+	 * Generates a form using the stored controller and model.
+	 *
+	 * @param string $name
+	 * @return Form $form
+	 */
+	public function getForm($name = 'Name') {
+		$record = $this->getRecord();
+		$fields = $this->getFormFields();
+		$actions = $this->getFormActions();
+		$validator = $this->getFormValidator();
+		$form = Form::create($this->controller, $name, $fields, $actions, $validator);
+
+		// Extend form
+		$this->extendAll('updateForm', $form);
+
+		// Load into form
+		$form->loadDataFrom($record);
+
+		return $form;
+	}
+
+	/**
+	 * Build field list for this form
+	 *
+	 * @return FieldList
+	 */
+	public function getFormFields() {
+		// Fall back to standard "getCMSFields" which itself uses the FormScaffolder as a fallback
+		$record = $this->getRecord();
+		$fields = $record->getCMSFields();
+		$this->extendAll('updateFormFields', $fields);
+		return $fields;
+	}
+
+	/**
+	 * Build list of actions for this form
+	 *
+	 * @return FieldList
+	 */
+	public function getFormActions() {
+		// by default no actions, it's a bit unpredictable
+		$record = $this->getRecord();
+
+		// Support legacy behaviour
+		$actions = $record->getCMSActions();
+		if ($actions->count()) {
+			Deprecation::notice('5.0', 'Implement updateFormActions() instead of providing getCMSActions()');
+		}
+
+		// Extend actions
+		$this->extendAll('updateFormActions', $actions);
+		return $actions;
+	}
+
+	/**
+	 * @return Controller
+	 */
+	public function getController() {
+		return $this->controller;
+	}
+
+	/**
+	 * @param Controller $controller
+	 * @return $this
+	 */
+	public function setController(Controller $controller) {
+		$this->controller = $controller;
+		return $this;
+	}
+
+	/**
+	 * @return DataObject
+	 */
+	public function getRecord() {
+		return $this->record;
+	}
+
+	/**
+	 * @param DataObject $record
+	 * @return $this
+	 */
+	public function setRecord(DataObject $record) {
+		$this->record = $record;
+		return $this;
+	}
+
+	/**
+	 * @return Validator|null
+	 */
+	public function getFormValidator() {
+		$validator = $this->validator;
+		$record = $this->getRecord();
+
+		// Support legacy behaviour
+		if (!$validator && $record->hasMethod('getCMSValidator')) {
+			Deprecation::notice('5.0', 'Implement updateFormValidator() instead of providing getCMSValidator()');
```

Comment:
Might be a bit too early to throw deprecation notices if we're not sure where this API goes?